exist-core/src/main/java/org/exist/xquery/value/OrderedValueSequence.java

```diff
@@ -328,59 +319,64 @@ public Entry(final Item item, final int position) throws XPathException {
+                                cmp = Constants.INFERIOR;
+                            }
+                        } else if (a == AtomicValue.EMPTY_VALUE && b != AtomicValue.EMPTY_VALUE) {
+                            if ((orderSpecs[i].getModifiers() & OrderSpec.EMPTY_LEAST) != 0) {
+                                cmp = Constants.INFERIOR;
+                            } else {
+                                cmp = Constants.SUPERIOR;
+                            }
+                        } else if (b == AtomicValue.EMPTY_VALUE && a != AtomicValue.EMPTY_VALUE) {
+                            if ((orderSpecs[i].getModifiers() & OrderSpec.EMPTY_LEAST) != 0) {
+                                cmp = Constants.SUPERIOR;
+                            } else {
+                                cmp = Constants.INFERIOR;
+                            }
                         } else {
-                            cmp = Constants.INFERIOR;
+                            cmp = a.compareTo(orderSpecs[i].getCollator(), b);
                         }
-                    } else if (a == AtomicValue.EMPTY_VALUE && b != AtomicValue.EMPTY_VALUE) {
-                        if ((orderSpecs[i].getModifiers() & OrderSpec.EMPTY_LEAST) != 0) {
-                            cmp = Constants.INFERIOR;
-                        } else {
-                            cmp = Constants.SUPERIOR;
+                        if ((orderSpecs[i].getModifiers() & OrderSpec.DESCENDING_ORDER) != 0) {
+                            cmp = cmp * -1;
                         }
-                    } else if (b == AtomicValue.EMPTY_VALUE && a != AtomicValue.EMPTY_VALUE) {
-                        if ((orderSpecs[i].getModifiers() & OrderSpec.EMPTY_LEAST) != 0) {
-                            cmp = Constants.SUPERIOR;
-                        } else {
-                            cmp = Constants.INFERIOR;
+                        if (cmp != Constants.EQUAL) {
+                            break;
                         }
-                    } else {
-                        cmp = a.compareTo(orderSpecs[i].getCollator(), b);
-                    }
-                    if ((orderSpecs[i].getModifiers() & OrderSpec.DESCENDING_ORDER) != 0) {
-                        cmp = cmp * -1;
-                    }
-                    if (cmp != Constants.EQUAL) {
-                        break;
+                    } catch (final XPathException e) {
                     }
-                } catch (final XPathException e) {
                 }
+                // if the sort keys are equal, we need to order by the original position in the result sequence
+                if (cmp == Constants.EQUAL) {
+                    cmp = (pos > other.pos ? Constants.SUPERIOR : (pos == other.pos ? Constants.EQUAL : Constants.INFERIOR));
+                }
+                return cmp;
+            } catch (RuntimeException e) {
+                e.printStackTrace();
+                throw e;
```

Comment:
no reporting with logger?

Reply:
Oops! That wasn't meant to make it into the PR. I have removed it now.